File: kotlin/protocol/src/main/kotlin/org/hpb/protocol/Assignment.kt
package org.hpb.protocol

import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.hpb.engine.hexBytes
import org.hpb.engine.nostr.Events
import org.hpb.engine.nostr.Nip44
import org.hpb.engine.nostr.NostrEvent
import org.hpb.protocol.Pj.a
import org.hpb.protocol.Pj.l
import org.hpb.protocol.Pj.s
import org.hpb.protocol.Pj.sOrNull

data class Claim(
    val offerEventId: String,
    val escrowId: String,
    val payoutAddress: String,
    val attestationEventIds: List<String>,
)

data class Grant(
    val claimEventId: String,
    val escrowId: String,
    val granted: Boolean,
    val taskKeys: List<String>,
    val expiresAt: Long,
    val reason: String? = null,
)

data class Answer(val taskKey: String, val answer: String)

data class Submission(val grantEventId: String, val escrowId: String, val answers: List<Answer>)

/** Claim/grant/resign/submission codecs. The grant is the countersignature
 *  that resolves claim races: only the offer author's grants allocate work. */
object Assignments {
    fun claim(privkey: ByteArray, launcher: String, c: Claim, createdAt: Long): NostrEvent =
        Events.sign(
            privkey, ProtocolKinds.CLAIM,
            tags = listOf(
                listOf("e", c.offerEventId),
                listOf("x", c.escrowId),
                listOf("p", launcher),
            ),
            content = Pj.obj(
                "v" to Pj.num(ProtocolKinds.VERSION),
                "payout_address" to Pj.str(c.payoutAddress),
                "attestations" to Pj.arr(c.attestationEventIds.map(Pj::str)),
            ).toString(),
            createdAt = createdAt,
        )

    fun parseClaim(event: NostrEvent): Claim {
        require(event.kind == ProtocolKinds.CLAIM) { "not a claim" }
        val content = Pj.parse(event.content)
        return Claim(
            offerEventId = requireNotNull(event.tagValue("e")),
            escrowId = requireNotNull(event.tagValue("x")),
            payoutAddress = content.s("payout_address"),
            attestationEventIds = content.a("attestations").map { it.jsonPrimitive.content },
        )
    }

    fun grant(privkey: ByteArray, worker: String, g: Grant, createdAt: Long): NostrEvent =
        Events.sign(
            privkey, ProtocolKinds.GRANT,
            tags = listOf(
                listOf("e", g.claimEventId),
                listOf("x", g.escrowId),
                listOf("p", worker),
            ),
            content = Pj.obj(
                "v" to Pj.num(ProtocolKinds.VERSION),
                "status" to Pj.str(if (g.granted) "granted" else "rejected"),
                "task_keys" to Pj.arr(g.taskKeys.map(Pj::str)),
                "expires_at" to Pj.num(g.expiresAt),
                "reason" to g.reason?.let(Pj::str),
            ).toString(),
            createdAt = createdAt,
        )

    fun parseGrant(event: NostrEvent): Grant {
        require(event.kind == ProtocolKinds.GRANT) { "not a grant" }
        val content = Pj.parse(event.content)
        return Grant(
            claimEventId = requireNotNull(event.tagValue("e")),
            escrowId = requireNotNull(event.tagValue("x")),
            granted = content.s("status") == "granted",
            taskKeys = content.a("task_keys").map { it.jsonPrimitive.content },
            expiresAt = content.l("expires_at"),
            reason = content.sOrNull("reason"),
        )
    }

    fun resign(
        privkey: ByteArray,
        launcher: String,
        grantEventId: String,
        escrowId: String,
        createdAt: Long,
    ): NostrEvent =
        Events.sign(
            privkey, ProtocolKinds.RESIGN,
            tags = listOf(
                listOf("e", grantEventId),
                listOf("x", escrowId),
                listOf("p", launcher),
            ),
            content = Pj.obj("v" to Pj.num(ProtocolKinds.VERSION)).toString(),
            createdAt = createdAt,
        )

    /** Answers are NIP-44 encrypted to the validator (copy-protection AND
     *  commitment). The nonce is random in production; fixed-nonce override
     *  exists ONLY so the vector corpus is byte-stable. */
    fun submission(
        privkey: ByteArray,
        validatorPubkey: String,
        s: Submission,
        createdAt: Long,
        nonce: ByteArray = randomNonce(),
    ): NostrEvent {
        val plaintext = Pj.obj(
            "v" to Pj.num(ProtocolKinds.VERSION),
            "answers" to Pj.arr(
                s.answers.map {
                    Pj.obj("task_key" to Pj.str(it.taskKey), "answer" to Pj.str(it.answer))
                },
            ),
        ).toString()
        val key = Nip44.conversationKey(privkey, validatorPubkey.hexBytes())
        return Events.sign(
            privkey, ProtocolKinds.SUBMISSION,
            tags = listOf(
                listOf("e", s.grantEventId),
                listOf("x", s.escrowId),
                listOf("p", validatorPubkey),
            ),
            content = Nip44.encrypt(plaintext, key, nonce),
            createdAt = createdAt,
        )
    }

    /** The validator decrypts with its key + the worker's pubkey. */
    fun decryptSubmission(event: NostrEvent, validatorPrivkey: ByteArray): List<Answer> {
        val key = Nip44.conversationKey(validatorPrivkey, event.pubkey.hexBytes())
        val content = Pj.parse(Nip44.decrypt(event.content, key))
        return content.a("answers").map {
            Answer(it.jsonObject.s("task_key"), it.jsonObject.s("answer"))
        }
    }

    private fun randomNonce(): ByteArray =
        ByteArray(32).also { java.security.SecureRandom().nextBytes(it) }
}

File: ios/HpbCore/Sources/HpbCore/Assignment.swift
import Foundation

public struct Claim {
    public let offerEventId: String
    public let escrowId: String
    public let payoutAddress: String
    public let attestationEventIds: [String]

    public init(offerEventId: String, escrowId: String, payoutAddress: String, attestationEventIds: [String]) {
        self.offerEventId = offerEventId
        self.escrowId = escrowId
        self.payoutAddress = payoutAddress
        self.attestationEventIds = attestationEventIds
    }
}

public struct Grant {
    public let claimEventId: String
    public let escrowId: String
    public let granted: Bool
    public let taskKeys: [String]
    public let expiresAt: Int64
    public let reason: String?

    public init(
        claimEventId: String, escrowId: String, granted: Bool,
        taskKeys: [String], expiresAt: Int64, reason: String? = nil
    ) {
        self.claimEventId = claimEventId
        self.escrowId = escrowId
        self.granted = granted
        self.taskKeys = taskKeys
        self.expiresAt = expiresAt
        self.reason = reason
    }
}

public struct Answer {
    public let taskKey: String
    public let answer: String

    public init(taskKey: String, answer: String) {
        self.taskKey = taskKey
        self.answer = answer
    }
}

public struct Submission {
    public let grantEventId: String
    public let escrowId: String
    public let answers: [Answer]

    public init(grantEventId: String, escrowId: String, answers: [Answer]) {
        self.grantEventId = grantEventId
        self.escrowId = escrowId
        self.answers = answers
    }
}

/// Claim/grant/resign/submission codecs. The grant is the countersignature
/// that resolves claim races: only the offer author's grants allocate work.
public enum Assignments {
    public static func claim(
        _ privkey: [UInt8], launcher: String, _ c: Claim, createdAt: Int64
    ) throws -> NostrEvent {
        try Events.sign(
            privkey, kind: ProtocolKinds.claim,
            tags: [
                ["e", c.offerEventId],
                ["x", c.escrowId],
                ["p", launcher],
            ],
            content: Cj.write(Cj.obj([
                ("v", Cj.num(ProtocolKinds.version)),
                ("payout_address", .str(c.payoutAddress)),
                ("attestations", .arr(c.attestationEventIds.map(J.str))),
            ])),
            createdAt: createdAt
        )
    }

    public static func parseClaim(_ event: NostrEvent) throws -> Claim {
        guard event.kind == ProtocolKinds.claim else { throw HpbError.invalid("not a claim") }
        let content = try Cj.parse(event.content)
        guard let offerEventId = event.tagValue("e"), let escrowId = event.tagValue("x") else {
            throw HpbError.invalid("claim missing tags")
        }
        return Claim(
            offerEventId: offerEventId,
            escrowId: escrowId,
            payoutAddress: try content.s("payout_address"),
            attestationEventIds: try content.a("attestations").compactMap(\.stringValue)
        )
    }

    public static func grant(
        _ privkey: [UInt8], worker: String, _ g: Grant, createdAt: Int64
    ) throws -> NostrEvent {
        try Events.sign(
            privkey, kind: ProtocolKinds.grant,
            tags: [
                ["e", g.claimEventId],
                ["x", g.escrowId],
                ["p", worker],
            ],
            content: Cj.write(Cj.obj([
                ("v", Cj.num(ProtocolKinds.version)),
                ("status", .str(g.granted ? "granted" : "rejected")),
                ("task_keys", .arr(g.taskKeys.map(J.str))),
                ("expires_at", Cj.num(g.expiresAt)),
                ("reason", g.reason.map(J.str)),
            ])),
            createdAt: createdAt
        )
    }

    public static func parseGrant(_ event: NostrEvent) throws -> Grant {
        guard event.kind == ProtocolKinds.grant else { throw HpbError.invalid("not a grant") }
        let content = try Cj.parse(event.content)
        guard let claimEventId = event.tagValue("e"), let escrowId = event.tagValue("x") else {
            throw HpbError.invalid("grant missing tags")
        }
        return Grant(
            claimEventId: claimEventId,
            escrowId: escrowId,
            granted: try content.s("status") == "granted",
            taskKeys: try content.a("task_keys").compactMap(\.stringValue),
            expiresAt: try content.l("expires_at"),
            reason: content.sOrNull("reason")
        )
    }

    public static func resign(
        _ privkey: [UInt8], launcher: String, grantEventId: String, escrowId: String, createdAt: Int64
    ) throws -> NostrEvent {
        try Events.sign(
            privkey, kind: ProtocolKinds.resign,
            tags: [
                ["e", grantEventId],
                ["x", escrowId],
                ["p", launcher],
            ],
            content: Cj.write(Cj.obj([("v", Cj.num(ProtocolKinds.version))])),
            createdAt: createdAt
        )
    }

    /// Answers are NIP-44 encrypted to the validator (copy-protection AND
    /// commitment). The nonce is random in production; the fixed-nonce
    /// override exists ONLY so the vector corpus is byte-stable.
    public static func submission(
        _ privkey: [UInt8], validatorPubkey: String, _ s: Submission,
        createdAt: Int64, nonce: [UInt8]? = nil
    ) throws -> NostrEvent {
        let plaintext = Cj.write(Cj.obj([
            ("v", Cj.num(ProtocolKinds.version)),
            ("answers", .arr(s.answers.map {
                Cj.obj([("task_key", .str($0.taskKey)), ("answer", .str($0.answer))])
            })),
        ]))
        let key = try Nip44.conversationKey(privkey, validatorPubkey.hexBytes())
        return try Events.sign(
            privkey, kind: ProtocolKinds.submission,
            tags: [
                ["e", s.grantEventId],
                ["x", s.escrowId],
                ["p", validatorPubkey],
            ],
            content: try Nip44.encrypt(plaintext, key, nonce ?? randomNonce()),
            createdAt: createdAt
        )
    }

    private static func randomNonce() -> [UInt8] {
        (0..<32).map { _ in UInt8.random(in: 0...255) }
    }
}
